Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,989 @@
+    (x : E) : ∃ (n : ℕ) (y : F), minpoly F x = X ^ q ^ n - C y :=
+  (isPurelyInseparable_iff_minpoly_eq_X_pow_sub_C F q).1 ‹_› x
+
+/-- A field extension `E / F` of exponential characteristic `q` is purely inseparable
+if and only if for every element `x` of `E`, the minimal polynomial of `x` over `F` is of form
+`(X - x) ^ (q ^ n)` for some natural number `n`. -/
+theorem isPurelyInseparable_iff_minpoly_eq_X_sub_C_pow (q : ℕ) [hF : ExpChar F q] :
+    IsPurelyInseparable F E ↔
+    ∀ x : E, ∃ n : ℕ, (minpoly F x).map (algebraMap F E) = (X - C x) ^ q ^ n := by
+  simp_rw [isPurelyInseparable_iff_natSepDegree_eq_one,
+    minpoly.natSepDegree_eq_one_iff_eq_X_sub_C_pow q]
+
+theorem IsPurelyInseparable.minpoly_eq_X_sub_C_pow (q : ℕ) [ExpChar F q] [IsPurelyInseparable F E]
+    (x : E) : ∃ n : ℕ, (minpoly F x).map (algebraMap F E) = (X - C x) ^ q ^ n :=
+  (isPurelyInseparable_iff_minpoly_eq_X_sub_C_pow F q).1 ‹_› x
+
+variable (E)
+
+-- TODO: remove `halg` assumption
+variable {F E} in
+/-- If an algebraic extension has finite separable degree one, then it is purely inseparable. -/
+theorem isPurelyInseparable_of_finSepDegree_eq_one (halg : Algebra.IsAlgebraic F E)
+    (hdeg : finSepDegree F E = 1) : IsPurelyInseparable F E := by
+  rw [isPurelyInseparable_iff]
+  refine fun x ↦ ⟨(halg x).isIntegral, fun hsep ↦ ?_⟩
+  have := finSepDegree_mul_finSepDegree_of_isAlgebraic F F⟮x⟯ E <| halg.tower_top F⟮x⟯
+  rw [hdeg, mul_eq_one, (finSepDegree_adjoin_simple_eq_finrank_iff F E x (halg x)).2 hsep,
+    IntermediateField.finrank_eq_one_iff] at this
+  simpa only [this.1] using mem_adjoin_simple_self F x
+
+/-- If `E / F` is purely inseparable, then for any reduced ring `L`, the map `(E →+* L) → (F →+* L)`
+induced by `algebraMap F E` is injective. In particular, a purely inseparable field extension
+is an epimorphism in the category of fields. -/
+theorem IsPurelyInseparable.injective_comp_algebraMap [IsPurelyInseparable F E]
+    (L : Type w) [CommRing L] [IsReduced L] :
+    Function.Injective fun f : E →+* L ↦ f.comp (algebraMap F E) := fun f g heq ↦ by
+  ext x
+  obtain ⟨q, hF⟩ := ExpChar.exists F
+  obtain ⟨n, y, h⟩ := IsPurelyInseparable.pow_mem F q x
+  replace heq := congr($heq y)
+  simp_rw [RingHom.comp_apply, h] at heq
+  cases hF
+  · rwa [one_pow, pow_one] at heq
+  nontriviality L
+  haveI := charP_of_injective_ringHom (f.comp (algebraMap F E)).injective q
+  haveI := Fact.mk ‹q.Prime›
+  simp_rw [map_pow, ← iterate_frobenius] at heq
+  exact (frobenius_inj L q).iterate n heq
+
+instance instSubsingletonAlgHomOfIsPurelyInseparable [IsPurelyInseparable F E] (L : Type w)
+    [CommRing L] [IsReduced L] [Algebra F L] : Subsingleton (E →ₐ[F] L) where
+  allEq f g := AlgHom.coe_ringHom_injective <|
+    IsPurelyInseparable.injective_comp_algebraMap F E L (by simp_rw [AlgHom.comp_algebraMap])
+
+instance instSubsingletonEmbOfIsPurelyInseparable [IsPurelyInseparable F E] :
+    Subsingleton (Emb F E) := by rw [Emb]; infer_instance
```

Comment:
I realize that we should upgrade [Field.instNonemptyEmb](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SeparableDegree.html#Field.instNonemptyEmb) to `Inhabited` and then this instance could be `Unique (Emb F E)` and we could use [Nat.card_unique](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Cardinal/Finite.html#Nat.card_unique) in the next proof.